File: javascript/linked-list/linked-list.js
export class LinkedList {
  constructor() {
    this.head = null
    this.tail = null
  }

  count(node = this.head, length = 0) {
    if (node) {
      return this.count(node.next, length + 1)
    }

    return length
  }

  push(value) {
    const node = this.insertAfterNode(this.tail, value)
    if (!this.head) {
      this.head = node
    }
    this.tail = node
  }

  insertAfterNode(node, value) {
    const newNode = { value, previos: node, next: null }
    if (node) {
      node.next = newNode
    }
    return newNode
  }

  unshift(value) {
    const node = this.insertBeforeNode(this.head, value)
    if (!this.tail) {
      this.tail = node
    }
    this.head = node
  }

  insertBeforeNode(node, value) {
    const newNode = { value, previos: null, next: node }
    if (node) {
      node.previos = newNode
    }
    return newNode
  }

  pop() {
    const { value } = this.tail
    this.deleteNode(this.tail)
    return value
  }

  shift() {
    const { value } = this.head
    this.deleteNode(this.head)
    return value
  }

  delete(value, node = this.head) {
    if (!node) {
      return
    }

    if (node.value === value) {
      this.deleteNode(node)
    } else {
      this.delete(value, node.next)
    }
  }

  deleteNode(node) {
    if (node === this.head && node === this.tail) {
      this.head = null
      this.tail = null
    }
    else if (node === this.head) {
      this.head = node.next
      this.head.previos = null
    } else if (node === this.tail) {
      this.tail = node.previos
      this.tail.next = null
    } else {
      node.previos.next = node.next
      node.next.previos = node.previos
    }
  }
}

File: ruby/linked-list/linked_list.rb
class Node
  attr_reader :value
  attr_accessor :previous_node, :next_node

  def initialize(value: nil, previous_node: nil, next_node: nil)
    @value = value
    @previous_node = previous_node
    @next_node = next_node
  end
end

class Deque
  def push(value)
    node = new_tail_node(value)
    return add_to_empty(node) if empty?

    tail.next_node = node
    self.tail = node
  end

  def pop
    return remove_last_node.value if head == tail

    value = tail.value
    self.tail = tail.previous_node
    value
  end

  def shift
    return remove_last_node.value if head == tail

    value = head.value
    self.head = head.next_node
    value
  end

  def unshift(value)
    node = new_head_node(value)
    return add_to_empty(node) if empty?

    head.previous_node = node
    self.head = node
  end

  private

  attr_accessor :head, :tail

  def new_tail_node(value)
    Node.new value: value, previous_node: tail
  end

  def new_head_node(value)
    Node.new value: value, next_node: head
  end

  def empty?
    head.nil? && tail.nil?
  end

  def add_to_empty(node)
    self.head = node
    self.tail = node
    node
  end

  def remove_last_node
    head.tap do
      self.head = nil
      self.tail = nil
    end
  end
end
